storage/namespace.go

```diff
@@ -446,14 +446,61 @@ func (n *dbNamespace) Flush(
 		}
 	}
 
-	// if nil, succeeded
+	res := multiErr.FinalError()
+	n.metrics.flush.ReportSuccessOrError(res, n.nowFn().Sub(callStart))
+	return res
+}
 
-	if res := multiErr.FinalError(); res != nil {
-		n.metrics.flush.ReportError(n.nowFn().Sub(callStart))
-		return res
+func (n *dbNamespace) FlushState(blockStart time.Time) fileOpState {
+	// NB(r): The logic below is not extensively clear just by reading the
+	// code. Essentially if all are success, we return success, if any
+	// are failed then we return failed with the minimum num failures to
+	// give all shards the chance of max retries - otherwise if any in progress
+	// we return in progress and if any is not started  we return not started.
+	// This is implemented the way it is to avoid iterating over all shards
+	// and reading/querying their states multiple times.
+	result := fileOpState{
+		Status:      fileOpNotStarted,
+		NumFailures: 0,
+	}
+	first := false
+	anyNumFailures := false
+	n.RLock()
+	for _, shard := range n.shards {
+		if shard == nil {
+			continue
+		}
+		state := shard.FlushState(blockStart)
+		switch state.Status {
+		case fileOpSuccess:
+			// All have to be success to report success so if it's the first then
+			// set it to success and if none others flag it otherwise we'll report
+			// it as success
+			if first {
+				result.Status = fileOpSuccess
+			}
+		case fileOpFailed:
+			// Always appear as failed if any failed
+			result.Status = fileOpFailed
+		case fileOpNotStarted, fileOpInProgress:
+			// If any in progress or any not started then make it appear as
+			// namespace has the same status
+			return state
+		}
+		if !anyNumFailures && state.NumFailures > 0 {
+			anyNumFailures = true
+			result.NumFailures = state.NumFailures
+		} else if state.NumFailures < result.NumFailures {
+			// Prefer reporting we have less failures, as this will determine
+			// if a retry is attempted or not - this means max retries is not
```

Comment:
seems the usage of MaxRetries() no longer means "max retries" now? should probably create a new field/ rename in options? 

Also does this mean it could retry forever if it keeps failing to flush?